### PR TITLE
[fix](function) resubmit-fix AES/SM3/SM4 encrypt/ decrypt algorithm initialization vector bug

### PR DESCRIPTION
**Pull request by xinyiZzz**
# Proposed changes

Issue Number: close #xxx

## Problem summary

The first fix: #17420, there is a compatibility problem,
So add aes/sm4 V2 function to support correct behavior.
The original aes/sm4 still maintains the previous behavior.

——————————————
1. ECB algorithm, `block_encryption_mode` does not take effect, it only takes effect when init vector is provided.
Solved: 192/256 supports calculation without init vector

2. For other algorithms, an error should be reported when there is no init vector

Initialization Vector. The default value for the block_encryption_mode system variable is aes-128-ecb, or ECB mode, which does not require an initialization vector. The alternative permitted block encryption modes CBC, CFB1, CFB8, CFB128, and OFB all require an initialization vector.

Reference: https://dev.mysql.com/doc/refman/8.0/en/encryption-functions.html#function_aes-decrypt

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xinyiZzz** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.33 seconds
 stream load tsv:          457 seconds loaded 74807831229 Bytes, about 156 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          73 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230322163025_clickbench_pr_118870.html

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java [hunk 1]
```diff
@@ -992,31 +1012,76 @@ private void analyzeBuiltinAggFunction(Analyzer analyzer) throws AnalysisExcepti
             if (ConnectContext.get() != null) {
                 blockEncryptionMode = ConnectContext.get().getSessionVariable().getBlockEncryptionMode();
                 if (fnName.getFunction().equalsIgnoreCase("aes_decrypt")
-                        || fnName.getFunction().equalsIgnoreCase("aes_encrypt")) {
+                        || fnName.getFunction().equalsIgnoreCase("aes_encrypt")
+                        || fnName.getFunction().equalsIgnoreCase("aes_decrypt_v2")
+                        || fnName.getFunction().equalsIgnoreCase("aes_encrypt_v2")) {
```
Comment (zhangstar333) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java:1017]: maybe should add some desc to explain old function and v2 function usage in docs

Comment (xinyiZzz) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/FunctionCallExpr.java:1017]: done

**Comment by xinyiZzz** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java [hunk 2]
```diff
@@ -574,6 +580,8 @@ public class BuiltinScalarFunctions implements FunctionHelper {
             scalar(Sm3sum.class, "sm3sum"),
             scalar(Sm4Decrypt.class, "sm4_decrypt"),
             scalar(Sm4Encrypt.class, "sm4_encrypt"),
+            scalar(Sm4DecryptV2.class, "sm4_decrypt_v2"),
+            scalar(Sm4EncryptV2.class, "sm4_encrypt_v2"),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java:584]: please add function in the order of alphabet

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/AesDecryptV2.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    public AesDecryptV2 withChildren(List<Expression> children) {
+        Preconditions.checkArgument(children.size() >= 2 && children.size() <= 4);
+        if (children.size() == 2) {
+            return new AesDecryptV2(children.get(0), children.get(1));
+        } else if (children().size() == 3) {
+            return new AesDecryptV2(children.get(0), children.get(1), children.get(2));
+        } else {
+            if (!(children.get(3) instanceof StringLiteral)) {
+                throw new AnalysisException("the 4th parameter should be string literal: " + this.toSql());
+            }
```
[the remark this answers is not in the document]
Comment (xinyiZzz) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/AesDecryptV2.java:68]: done

File: fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java (hunk 2 above)
Comment (xinyiZzz) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java:584]: done

**Comment by xinyiZzz** [comment]
run buildall

**Review by zhangstar333 — DISMISSED** [review]

**Comment by xinyiZzz** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/Sm4DecryptV2.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    public Sm4DecryptV2 withChildren(List<Expression> children) {
+        Preconditions.checkArgument(children.size() >= 2 && children.size() <= 4);
+        if (children.size() == 2) {
+            return new Sm4DecryptV2(children.get(0), children.get(1));
+        } else if (children().size() == 3) {
+            return new Sm4DecryptV2(children.get(0), children.get(1), children.get(2));
+        } else {
+            if (!(children.get(3) instanceof StringLiteral)) {
+                throw new AnalysisException("the 4th parameter should be string literal: " + this.toSql());
+            }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/Sm4DecryptV2.java:63]: remaining check, please check all withChildren

File: fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java [hunk 5]
```diff
@@ -338,6 +342,8 @@ public class BuiltinScalarFunctions implements FunctionHelper {
             scalar(Acos.class, "acos"),
             scalar(AesDecrypt.class, "aes_decrypt"),
             scalar(AesEncrypt.class, "aes_encrypt"),
+            scalar(AesDecryptV2.class, "aes_decrypt_v2"),
+            scalar(AesEncryptV2.class, "aes_encrypt_v2"),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java:346]:
```suggestion
            scalar(AesDecrypt.class, "aes_decrypt"),
            scalar(AesDecryptV2.class, "aes_decrypt_v2"),
            scalar(AesEncrypt.class, "aes_encrypt"),
            scalar(AesEncryptV2.class, "aes_encrypt_v2"),
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/Sm4DecryptV2.java (hunk 4 above)
Comment (xinyiZzz) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/Sm4DecryptV2.java:63]: thx~

**Comment by xinyiZzz** [comment]
run buildall

**Comment by xinyiZzz** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**